### DISCUSSION TITLE
Logging not consistent, and rarely might log to a file that I did not specify

### DISCUSSION THREAD

**Discussion by amvarunn** [strongswan/strongswan#1546] [Q&A] [answered]
strongswan version : 5.9.6

Kindly help us with some inputs for the below problem.

We have an issue where the logs sometimes are written to a file that we did not specify, even though we can see filelogger having the correct path

/etc/strongswan/strongswan.conf

```
charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
}

include strongswan.d/*.conf

```

/etc/strongswan/strongswan.d/cmp_strongswan.conf
```
charon {
        load_modular = yes
        make_before_break = yes
        retransmit_base = 1
        retransmit_timeout = 15.0
        retransmit_tries = 2
        plugins {
                duplicheck {
                    enable = no
                }
        }

# two defined file loggers
    filelog {
        charon {
            # add a timestamp prefix
            time_format = %a %b %0e %T %Y
            # prepend connection name, simplifies grepping
            ike_name = yes
            # Append existing files
            append = yes
            # increase default loglevel for all daemon subsystems
            default = 0
            ike = 1
            cfg = 1
            enc = 0
            # flush each line to disk
            flush_line = yes
        }
        stderr {
            # more detailed loglevel for a specific subsystem, overriding the
            # default loglevel.
        }
   }
}
include /data/etc/strongswan/charon-logfile.conf
```

/data/etc/strongswan/charon-logfile.conf

```
charon {
    filelog {
        charon {
            # path to the log file, specify this as section name in versions prior to 5.7.0
            path = /new_dir/charon.identifier.log
        }
   }
}
```

The 'identifier' above is derived from a machine identifier env variable, and we can see that it is appropriately populated  when charon-logfile.conf is generated. The file itself is generated by a script, and once generated(at reboot), it will not change.

We have an issue where, very rarely, after logrotate(or swanctl --reload-settings) the logging starts to happen in a file named `/new_dir/charon` instead of ` /new_dir/charon.identifier.log`.
It goes back to logging in `/new_dir/charon.identifier.log` when we do a strongswan restart, but issue repeats on the next logrotate.

Our logrotate for reference
```
/new_dir/charon.*.log {
    su root root
    rotate 3
    size +1M
    notifempty
    missingok
    create 666 root root
    sharedscripts
    compress
    postrotate
        /usr/bin/killall -HUP /usr/sbin/charon-systemd
        /usr/bin/killall -HUP /usr/libexec/strongswan/charon
    endscript
}
```

The beginning of 'charon' log does not indicate any problem, there are a couple of lines followed by normal logging
```
Tue Feb 21 14:28:31 2023 13[CFG] loaded 0 entries for attr plugin configuration
Tue Feb 21 14:28:31 2023 13[CFG] loaded 0 RADIUS server configurations
```

Some additional info
```
Before strongswan restart(logging happening in file named charon)

[14:51:40] sys-1-N2:/new_dir # fuser /new_dir/charon
/new_dir/charon:        3321

[14:51:52] sys-1-N2:/new_dir # ps -p 3321
  PID TTY          TIME CMD
 3321 ?        00:00:26 charon
2:51

[14:52:37] sys-1-N2:/new_dir # fuser /new_dir/charon.identifier.log
/new_dir/charon.identifier.log:  1023

[14:52:44] sys-1-N2:/new_dir # ps -p 1023
  PID TTY          TIME CMD
 1023 ?        00:00:24 charon-systemd



After restart (Logging happens in correct file until next logrotate/swanctl --reload-settings)
[14:54:44] sys-1-N2:/new_dir # fuser /new_dir/charon

[14:55:02] sys-1-N2:/new_dir # fuser /new_dir/charon.identifier.log
/new_dir/charon.identifier.log:  1023 17902

[14:55:12] sys-1-N2:/new_dir # ps -p 17902
  PID TTY          TIME CMD
17902 ?        00:00:00 charon
```

This has happened more than once, however at this point, it is almost impossible to reproduce it willingly.
(Commenting the file path in filelog and doing swanctl --reload-settings reproduces it, but not in the same way, as in this case strongswan restart fails to get the correct name as expected)

Please help me understand, why most of the times this work and only fails sometime.
From the above, it looks like charon process may log either to specified file, or to a file named 'charon', why is this behavior happening even though we can see correct file path in the conf file(And if indeed the problem was with the conf file itself, why does strongswan restart log on to the correct file?).

Thank you

**Comment by tobiasbrunner** [comment, marked as answer]
If `path` is not configured in a subsection of `charon.filelog`, the section name itself (i.e. `charon` in your case) specifies the file/path for the log file (see [the documentation](https://docs.strongswan.org/docs/5.9/config/logging.html#_file_loggers)). If it's not an absolute path, the file will be created relative to the daemon's working directory.

Now the question is, why is `path` specified when the daemon is (re-)started but not when the config is reloaded afterwards. My guess is that there a switch to a different user after the config/plugins were loaded initially (search the log for a message like: `dropped capabilities, running as uid ..., gid ...`). And while the daemon still can open and read `/etc/strongswan/strongswan.d/cmp_strongswan.conf` when the config is reloaded, it fails to do so for `/data/etc/strongswan/charon-logfile.conf`. Please check the permissions of the directories and file in that path and make sure the daemon (i.e. its user and/or group) still can read/enter them after the user switch.

**Reply by amvarunn** [reply]
Thank you. 
By changing the permission of  `/data/etc/strongswan/charon-logfile.conf` we are now able to recreate this issue.



> If it's not an absolute path, the file will be created relative to the daemon's working directory.


Please provide some additional clarification regarding above.
The working directory of the deamon is not `/new_dir`. 
So should we be looking for another possible log file, generated elsewhere?

**Reply by tobiasbrunner** [reply]
Possibly, the configured file name/path (`path` or the section name) is just passed to `fopen()`.
